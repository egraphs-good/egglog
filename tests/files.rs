use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
