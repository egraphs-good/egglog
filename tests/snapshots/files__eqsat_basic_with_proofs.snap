---
source: tests/files.rs
expression: serialized
---
nodes:
  Repv484________-0:
    op: Repv484________
    children: []
    eclass: "261"
    cost: 1
  Ast*_i64_i64__-8490870082613382153:
    op: Ast*_i64_i64__
    children: []
    eclass: "164"
    cost: 1
  String-7467262353656187021:
    op: "\"(rewrite (Add a b) (Add b a))\""
    children: []
    eclass: String-7467262353656187021
    cost: 0
  Rule__-12805803976824527279:
    op: Rule__
    children:
      - Cons__-745149374760124569
      - String-7467262353656187021
    eclass: "46"
    cost: 1
  Rule__-2342299780272692938:
    op: Rule__
    children:
      - Cons__-3984406874221510308
      - String-7467262353656187021
    eclass: "54"
    cost: 1
  String-11676241252185100960:
    op: "\"(rewrite (Mul a (Add b c)) (Add (Mul a b) (Mul a c)))\""
    children: []
    eclass: String-11676241252185100960
    cost: 0
  Rule__-5721546683204446421:
    op: Rule__
    children:
      - Cons__-8588982171855026655
      - String-11676241252185100960
    eclass: "64"
    cost: 1
  Rule__-8563633762492217552:
    op: Rule__
    children:
      - Cons__-9243618720786002930
      - String-7467262353656187021
    eclass: "93"
    cost: 1
  Rule__-10635466700685145488:
    op: Rule__
    children:
      - Cons__-11315451658978930866
      - String-7467262353656187021
    eclass: "115"
    cost: 1
  Rule__-5357662616185599202:
    op: Rule__
    children:
      - Cons__-2369507322758629044
      - String-7467262353656187021
    eclass: "127"
    cost: 1
  String-1368580029154570717:
    op: "\"(rewrite (Mul (Num a) (Num b)) (Num (* a b)))\""
    children: []
    eclass: String-1368580029154570717
    cost: 0
  Rule__-15871410596505066269:
    op: Rule__
    children:
      - Cons__-8390451017274507649
      - String-1368580029154570717
    eclass: "155"
    cost: 1
  Rule__-11321138220942952796:
    op: Rule__
    children:
      - String-11676241252185100960
    eclass: "175"
    cost: 1
  Rule__-13566246915761015865:
    op: Rule__
    children:
      - String-7467262353656187021
    eclass: "241"
    cost: 1
  String-5696853317067581256:
    op: "\"(rule ((= expr1 expr2))\n      ((union v483_____ rule-proof))\n         )\""
    children: []
    eclass: String-5696853317067581256
    cost: 0
  Rule__-7726266060706311350:
    op: Rule__
    children:
      - String-5696853317067581256
    eclass: "258"
    cost: 1
  Cons__-745149374760124569:
    op: Cons__
    children: []
    eclass: "45"
    cost: 1
  Cons__-3984406874221510308:
    op: Cons__
    children: []
    eclass: "53"
    cost: 1
  Cons__-8588982171855026655:
    op: Cons__
    children: []
    eclass: "63"
    cost: 1
  Cons__-9243618720786002930:
    op: Cons__
    children: []
    eclass: "92"
    cost: 1
  Cons__-11315451658978930866:
    op: Cons__
    children: []
    eclass: "114"
    cost: 1
  Cons__-2369507322758629044:
    op: Cons__
    children: []
    eclass: "126"
    cost: 1
  Cons__-8179951341697187233:
    op: Cons__
    children: []
    eclass: "148"
    cost: 1
  Cons__-9200723339474947083:
    op: Cons__
    children:
      - Cons__-8179951341697187233
    eclass: "149"
    cost: 1
  Cons__-8390451017274507649:
    op: Cons__
    children: []
    eclass: "154"
    cost: 1
  Cons__-8030150736068464880:
    op: Cons__
    children: []
    eclass: "166"
    cost: 1
  i64-6755155689022739364:
    op: "20"
    children: []
    eclass: i64-6755155689022739364
    cost: 0
  currentAge-0:
    op: currentAge
    children: []
    eclass: i64-6755155689022739364
    cost: 1
  expr2-0:
    op: expr2
    children: []
    eclass: "26"
    cost: 9223372036854776000
  AstAdd_Math_Math__-1955564354691009820:
    op: AstAdd_Math_Math__
    children:
      - AstNum_i64__-14289738803621830331
    eclass: "23"
    cost: 1
  AstAdd_Math_Math__-15013121528888003876:
    op: AstAdd_Math_Math__
    children:
      - AstNum_i64__-10132733533534109046
    eclass: "42"
    cost: 1
  AstAdd_Math_Math__-7187812301797656813:
    op: AstAdd_Math_Math__
    children:
      - AstNum_i64__-10132733533534109046
    eclass: "88"
    cost: 1
  AstAdd_Math_Math__-1456766727393444901:
    op: AstAdd_Math_Math__
    children:
      - AstNum_i64__-14289738803621830331
    eclass: "122"
    cost: 1
  AstAdd_Math_Math__-5992595848691763978:
    op: AstAdd_Math_Math__
    children: []
    eclass: "110"
    cost: 1
  AstAdd_Math_Math__-5196680987275522562:
    op: AstAdd_Math_Math__
    children: []
    eclass: "208"
    cost: 1
  RepAdd_Math_Math__-3556106445369693066:
    op: RepAdd_Math_Math__
    children: []
    eclass: "25"
    cost: 1
  RepAdd_Math_Math__-10146167964587346060:
    op: RepAdd_Math_Math__
    children: []
    eclass: "59"
    cost: 1
  RepAdd_Math_Math__-14597414465048629250:
    op: RepAdd_Math_Math__
    children:
      - Mul-7930179513703693375
      - Mul-13853934148708888482
    eclass: "86"
    cost: 1
  RepAdd_Math_Math__-2110467172932255177:
    op: RepAdd_Math_Math__
    children:
      - Mul-13853934148708888482
      - Mul-7930179513703693375
    eclass: "120"
    cost: 1
  Mul-447570027545795062:
    op: Mul
    children: []
    eclass: "26"
    cost: 1
  Mul-7930179513703693375:
    op: Mul
    children: []
    eclass: "37"
    cost: 1
  Mul-13853934148708888482:
    op: Mul
    children: []
    eclass: "66"
    cost: 1
  AstNum_i64__-5871781006564002453:
    op: AstNum_i64__
    children: []
    eclass: "181"
    cost: 1
  AstNum_i64__-14289738803621830331:
    op: AstNum_i64__
    children: []
    eclass: "134"
    cost: 1
  AstNum_i64__-10132733533534109046:
    op: AstNum_i64__
    children: []
    eclass: "99"
    cost: 1
  AstNum_i64__-3741393243599718260:
    op: AstNum_i64__
    children:
      - Ast*_i64_i64__-8490870082613382153
    eclass: "163"
    cost: 1
root_eclasses: []
class_data:
  "261":
    type: TrmPrf__
  "164":
    type: Ast__
  "1":
    type: Ast__
  "15":
    type: Ast__
  "46":
    type: Proof__
  "45":
    type: ProofList__
  String-7467262353656187021:
    type: String
  "54":
    type: Proof__
  "53":
    type: ProofList__
  "64":
    type: Proof__
  "63":
    type: ProofList__
  String-11676241252185100960:
    type: String
  "93":
    type: Proof__
  "92":
    type: ProofList__
  "115":
    type: Proof__
  "114":
    type: ProofList__
  "127":
    type: Proof__
  "126":
    type: ProofList__
  "155":
    type: Proof__
  "154":
    type: ProofList__
  String-1368580029154570717:
    type: String
  "175":
    type: Proof__
  "174":
    type: ProofList__
  "241":
    type: Proof__
  "240":
    type: ProofList__
  "258":
    type: Proof__
  "263":
    type: ProofList__
  String-5696853317067581256:
    type: String
  "43":
    type: Proof__
  "0":
    type: ProofList__
  "24":
    type: Proof__
  "28":
    type: Proof__
  "62":
    type: ProofList__
  "89":
    type: Proof__
  "111":
    type: Proof__
  "123":
    type: Proof__
  "148":
    type: ProofList__
  "6":
    type: Proof__
  "149":
    type: ProofList__
  "20":
    type: Proof__
  "145":
    type: Proof__
  "153":
    type: ProofList__
  "166":
    type: ProofList__
  "165":
    type: Proof__
  i64-6755155689022739364:
    type: i64
  "26":
    type: Math
  "23":
    type: Ast__
  "133":
    type: Ast__
  "134":
    type: Ast__
  "42":
    type: Ast__
  "99":
    type: Ast__
  "100":
    type: Ast__
  "88":
    type: Ast__
  "122":
    type: Ast__
  "110":
    type: Ast__
  "248":
    type: Ast__
  "208":
    type: Ast__
  "25":
    type: TrmPrf__
  "11":
    type: Math
  "18":
    type: Math
  "59":
    type: TrmPrf__
  "86":
    type: TrmPrf__
  "37":
    type: Math
  "66":
    type: Math
  "120":
    type: TrmPrf__
  "4":
    type: Math
  "22":
    type: Math
  "181":
    type: Ast__
  "30":
    type: Ast__
  "163":
    type: Ast__

